tests/expose_test.go

```diff
@@ -81,34 +94,96 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 		Context("Expose ClusterIP service", func() {
 			const servicePort = "27017"
 			const serviceName = "cluster-ip-vmi"
-			It("[test_id:1531][label:masquerade_binding_connectivity]Should expose a Cluster IP service on a VMI and connect to it", func() {
+
+			var vmiExposeArgs []string
+
+			var jobCleanupFunc func() error
+			var serviceCleanupFunc func() error
+
+			BeforeEach(func() {
+				vmiExposeArgs = []string{
+					expose.COMMAND_EXPOSE,
+					"virtualmachineinstance", "--namespace", tcpVM.GetNamespace(), tcpVM.GetName(),
+					"--port", servicePort, "--name", serviceName,
+					"--target-port", strconv.Itoa(testPort),
+				}
+			})
+
+			BeforeEach(func() {
+				jobCleanupFunc = nil
+			})
+
+			BeforeEach(func() {
+				serviceCleanupFunc = nil
+			})
+
+			AfterEach(func() {
+				Expect(jobCleanupFunc).NotTo(BeNil(), "a successful test must have stored a way to delete the batchv1.Job entity")
+				Expect(jobCleanupFunc()).To(Succeed(), "should be able to delete the batchv1.Job entity")
+			})
+
+			AfterEach(func() {
+				Expect(serviceCleanupFunc).NotTo(BeNil(), "a successful test must have stored a way to delete the k8sv1.Service entity")
+				Expect(serviceCleanupFunc()).To(Succeed(), "should be able to delete the k8sv1.Service entity")
+			})
+
+			table.DescribeTable("[test_id:1531][label:masquerade_binding_connectivity]Should expose a Cluster IP service on a VMI and connect to it", func(ipFamily k8sv1.IPFamily) {
 				By("Exposing the service via virtctl command")
-				virtctl := tests.NewRepeatableVirtctlCommand(expose.COMMAND_EXPOSE, "virtualmachineinstance", "--namespace",
-					tcpVM.Namespace, tcpVM.Name, "--port", servicePort, "--name", serviceName, "--target-port", strconv.Itoa(testPort))
-				err := virtctl()
-				Expect(err).ToNot(HaveOccurred())
+				virtctl := tests.NewRepeatableVirtctlCommand(getVirtctlExposeCommand(vmiExposeArgs, ipFamily)...)
+				Expect(virtctl()).To(Succeed(), "should expose a service via `virtctl expose ...`")
 
 				By("Getting back the cluster IP given for the service")
 				svc, err := virtClient.CoreV1().Services(tcpVM.Namespace).Get(serviceName, k8smetav1.GetOptions{})
 				Expect(err).ToNot(HaveOccurred())
+				serviceCleanupFunc = cleanupService(serviceName, tcpVM.GetNamespace())
```

Comment:
Is it possiable to move this to the beforeEach() (or maybe the afterEach()?) ?  I think it is more suitable there cause it is not so related to the test itself, but more of an infra thing. maybe instead of :
```
BeforeEach(func() {
    serviceCleanupFunc = nil
})
```
you can: 
```
BeforeEach(func() {
    serviceCleanupFunc = cleanupService(serviceName, tcpVM.GetNamespace())
})

Reply:
Than maybe in the afterEach()?
why it's important to keep it in the test body?

Reply:
> Right. Then maybe in the afterEach()?

That would be possible, but I don't see what's the point. I'm creating the service in the test, and storing the pointer to it also on the test. Then,  the test tear-down just calls the cleanup function. 

What you say would require me to store a pointer to the service, and then I'd have to use this to delete it. What are the advantages ? What are you trying to prevent on the current approach ? 

> why it's important to keep it in the test body?

The order of the tests (as described in our functional test plan) is:
  - create VM
  - expose service from stopped VM
  - start VM
  - check connectivity

We are creating the tests on a table, having the IP family as the parameter. Accessing the IP family is only possible **in** the tests, not in the setup, nor tear-down of the test.

This limits us to expose the service *within* the tests. If we want to preserve what the test does (I think we should) we can create the VM on the test setup, but are forced to expose the service & start the VM in the test code.

Reply:
> > > Right. Then maybe in the afterEach()?
> > 
> > 
> > That would be possible, but I don't see what's the point. I'm creating the service in the test, and storing the pointer to the callback that deletes it also on the test. Then, the test tear-down just calls the cleanup function.
> > What you say would require me to store a pointer to the service, and then I'd have to use this to delete it. What are the advantages ? What are you trying to prevent on the current approach ?
> > > why it's important to keep it in the test body?
> > 
> > 
> > The order of the tests (as described in our functional test plan) is:
> > 
> > * create VM
> > * expose service from stopped VM
> > * start VM
> > * check connectivity
> > 
> > We are creating the tests on a table, having the IP family as the parameter. Accessing the IP family is only possible **in** the tests, not in the setup, nor tear-down of the test.
> > This limits us to expose the service _within_ the tests. If we want to preserve what the test does (I think we should) we can create the VM on the test setup, but are forced to expose the service & start the VM in the test code.
> 
> I think I get what I missed here- that it is a table and the afterEach won't run after each entry, so you had to use the callback. am I right?

What ? No, the table entries generate 2 tests, which have setup / tear-down.

> My point was to keep things simpler, avoid using callbacks, and keep the tests without teardown-related processes, but it is not that crucial I think.
> Thanks for explaining in such details.

Cleaning stuff up in the tear-down is the proper way to handle this.

From where I stand, it seems you don't like the callback; I'm OK with that, if you insist, I can change it to be a regular function call .

> 
> By the way, in order to run beforeEach() and afterEach() after each table entry it is possible to do something like suggested here:
> [onsi/ginkgo#378 (comment)](https://github.com/onsi/ginkgo/issues/378#issuecomment-501538323)
> but again, not crucial, and thanks for explaining.